stdlib/runtime/concat.gr

```diff
@@ -0,0 +1,64 @@
+/* grainc-flags --no-pervasives */
+module Concat
+
+include "runtime/unsafe/wasmi32"
+include "runtime/unsafe/memory"
+include "runtime/dataStructures"
+from DataStructures use { tagSimpleNumber, allocateArray }
+
+let rec append = (list1, list2) => {
+  match (list1) {
+    [] => list2,
+    [first, ...rest] => [first, ...append(rest, list2)],
+  }
+}
+
+@unsafe
+provide let listConcat = listsArray => {
+  let ptr = WasmI32.fromGrain(listsArray)
+  let length = WasmI32.load(ptr, 4n)
+  let mut result = []
+
+  for (let mut i = 0n; WasmI32.ltU(i, length); i = WasmI32.add(i, 1n)) {
+    let list = listsArray[tagSimpleNumber(i)]
+    result = append(result, list)
+  }
+  result
+}
+
+@unsafe
+let findLength = arr => {
+  let ptr = WasmI32.fromGrain(arr)
+  WasmI32.load(ptr, 4n)
+}
+
+@unsafe
+provide let arrayConcat = (arraysArray: Array<Array<a>>) => {
+  let numArrs = findLength(arraysArray)
+
+  let mut totLength = 0n
+  for (let mut i = 0n; WasmI32.ltU(i, numArrs); i = WasmI32.add(i, 1n)) {
+    let array = arraysArray[tagSimpleNumber(i)]
+    totLength = WasmI32.add(totLength, findLength(array))
+  }
+
+  let result = allocateArray(totLength)
+  let mut byteOffset = result
+  for (let mut i = 0n; WasmI32.ltU(i, numArrs); i = WasmI32.add(i, 1n)) {
+    let array = arraysArray[tagSimpleNumber(i)]
```

Comment:
I don't usually say this, but for the sake of runtime performance, it's likely worth it to directly load the values instead of using the syntax because of the unnecessary bounds checking.